File: Leetcode/dfs_recursion_backtracking/547_FriendCircles/Solution.py
class Solution(object):
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        def dfs(num):
            if M[num][num]==-1:
                return
            M[num][num] = -1
            for i in xrange(0, len(M[num])):
                if(M[num][i]==1):  dfs(i)
        res = 0
        for i in xrange(0, len(M)):
            if M[i][i]==1:
                dfs(i)
                res += 1
        return res

File: Leetcode/dfs_recursion_backtracking/547_FriendCircles/Solution.java
class Solution {
    public int findCircleNum(int[][] M) {
        int res = 0;
        for(int i = 0; i<M.length; ++i){
            if( M[i][i]==1 ){
                dfs(M, i);
                res += 1;
            }
        }
        return res;
    }
        private void dfs(int[][] M, int id){
            if(M[id][id]==-1)
                return;
            M[id][id] = -1;
            for(int i = 0; i<M[id].length; ++i){
                if( M[id][i]==1 )
                    dfs(M, i);
            }
            return;
        }
}
